File: dispatch/core/RemoteFunction/RemoteFunctionRunnable.java
package adamma.c4dhi.claid.RemoteFunction;


// The RemoteFunctionRunnable is the counterpart to the RemoteFunction.
// The RemoteFunctionRunnable is the actual function, which get's called remotely.
// E.g., it could be a function registered inside a Module. 

import java.lang.reflect.Method;

import java.util.ArrayList;

import adamma.c4dhi.claid.RemoteFunctionRequest;
import adamma.c4dhi.claid.RemoteFunctionReturn;
import adamma.c4dhi.claid.RemoteFunctionStatus;

import adamma.c4dhi.claid.Logger.Logger;
import adamma.c4dhi.claid.TypeMapping.DataType;
import adamma.c4dhi.claid.TypeMapping.Mutator;
import adamma.c4dhi.claid.TypeMapping.TypeMapping;

import adamma.c4dhi.claid.DataPackage;
import adamma.c4dhi.claid.ControlPackage;
import adamma.c4dhi.claid.CtrlType;
import adamma.c4dhi.claid.Blob;
import adamma.c4dhi.claid.Runtime;

import adamma.c4dhi.claid.RemoteFunctionIdentifier;

import adamma.c4dhi.claid.RemoteFunction.RemoteFunctionRunnableResult;
import adamma.c4dhi.claid.RemoteFunction.RemoteFunctionRunnable;

import com.google.protobuf.GeneratedMessageV3;

public class RemoteFunctionRunnable 
{
    private String functionName;
    private Class<?> returnType;
    private ArrayList<Class<?>> parameterTypes;

    RemoteFunctionRunnable(String functionName, Class<?> returnType, ArrayList<Class<?>> parameterTypes)
    {
        this.functionName = functionName;
        this.returnType = returnType;
        this.parameterTypes = parameterTypes;
    }

    DataPackage executeRemoteFunctionRequest(Object object, DataPackage rpcRequest)
    {
        RemoteFunctionRunnableResult status;

        RemoteFunctionRequest executionRequest = rpcRequest.getControlVal().getRemoteFunctionRequest();

        RemoteFunctionIdentifier remoteFunctionIdentifier = executionRequest.getRemoteFunctionIdentifier();

        int payloadsSize = executionRequest.getParameterPayloadsCount();

        if(payloadsSize!= this.parameterTypes.size())
        {
            Logger.logError("Failed to execute RemoteFunctionRunnable \"" + getFunctionSignature(remoteFunctionIdentifier, executionRequest) + "\". Number of parameters do not match. " +
            "Function expected " + parameterTypes.size() + " parameters, but was executed with " + payloadsSize);

            status = RemoteFunctionRunnableResult.makeFailedResult(RemoteFunctionStatus.FAILED_INVALID_NUMBER_OF_PARAMETERS);
            return makeRPCResponsePackage(status, rpcRequest);
        }

        ArrayList<Object> parameters = new ArrayList<>();
        for(int i = 0; i < payloadsSize; i++)
        {
            Class<?> dataType = this.parameterTypes.get(i);
            Mutator<?> mutator = TypeMapping.getMutator(new DataType(dataType));

            DataPackage.Builder tmpPackage = DataPackage.newBuilder();
            tmpPackage.setPayload(executionRequest.getParameterPayloads(i));

            Object data = mutator.getPackagePayload(tmpPackage.build());
            parameters.add(data);

            if(!parameters.get(i).getClass().equals(this.parameterTypes.get(i)))
            {
                Logger.logError("Failed to execute RemoteFunctionRunnable \"" + getFunctionSignature(remoteFunctionIdentifier, executionRequest) + "\". Parameter object " + i +
                " is of type \"" + parameters.get(i).getClass() + "\", but expected type \"" + parameterTypes.get(i).getSimpleName() + "\".");
                status = RemoteFunctionRunnableResult.makeFailedResult(RemoteFunctionStatus.FAILED_MISMATCHING_PARAMETERS);
                return makeRPCResponsePackage(status, rpcRequest);
            }
        }

        status = executeRemoteFunctionRequest(object, parameters);

        return makeRPCResponsePackage(status, rpcRequest);
    }

    RemoteFunctionRunnableResult executeRemoteFunctionRequest(Object object, ArrayList<Object> parameters)
    {
        try 
        {        
            Method method = RemoteFunctionRunnable.lookupMethod(object, functionName, parameterTypes);
            if(method == null)
            {
                return RemoteFunctionRunnableResult.makeFailedResult(RemoteFunctionStatus.FAILED_FUNCTION_NOT_FOUND_OR_FAILED_TO_EXECUTE);
            }

            // Convert the ArrayList to an array of Objects
            Object[] paramsArray = parameters.toArray();

            // Invoke the method on the instance with the parameters
            Object result = method.invoke(object, paramsArray);

            return RemoteFunctionRunnableResult.makeSuccessfulResult(result);
        } 
        catch (Exception e) 
        {
            e.printStackTrace();
            Logger.logError("Failed to execute RemoteFunctionRunnable. Got exception \"" + e.getMessage() + "\".");
            return RemoteFunctionRunnableResult.makeFailedResult(RemoteFunctionStatus.FAILED_FUNCTION_NOT_FOUND_OR_FAILED_TO_EXECUTE);
        }
    }

    public static Method lookupMethod(Object object, String functionName, ArrayList<Class<?>> parameterTypes)
    {
        Logger.logError("dbg 1");
        try 
        {        
            Class<?> myClass = object.getClass();
            Logger.logError("dbg 2");

            Class<?>[] parameterTypesArray = new Class<?>[parameterTypes.size()];
            parameterTypesArray = parameterTypes.toArray(parameterTypesArray);
            Logger.logError("dbg 3");

            Method method = myClass.getMethod(functionName, parameterTypesArray);        
            Logger.logError("dbg 4 " + method);

            return method;
        } 
        catch (Exception e) 
        {
            e.printStackTrace();
            Logger.logError("Failed to lookup function Method \"" + functionName + "\" of object of type \"" + object.getClass().getName() + "\". Got exception \"" + e.getMessage() + "\".");
            return null;
        }
    }

    public static boolean doesFunctionExist(Object object, String functionName, ArrayList<Class<?>> parameterTypes)
    {
        Logger.logError("Dbg 0 " + parameterTypes.size());
        return RemoteFunctionRunnable.lookupMethod(object, functionName, parameterTypes) != null;
    }

    public static boolean isDataTypeSupported(Class<?> type)
    {
        Mutator<?> mutator = TypeMapping.getMutator(new DataType(type));
        
        return mutator != null;
    }

    public String getFunctionSignature(RemoteFunctionIdentifier remoteFunctionIdentifier, RemoteFunctionRequest RemoteFunctionRequest)
    {
        String returnTypeName = returnType == null ? "void" : returnType.getSimpleName();

        boolean isRuntimeFunction = remoteFunctionIdentifier.hasRuntime();
    
        String parameterNames = this.parameterTypes.size() > 0 ? this.parameterTypes.get(0).getSimpleName() : "";

        for(int i = 1; i < this.parameterTypes.size(); i++)
        {
            parameterNames += ", " + this.parameterTypes.get(i).getSimpleName();
        }

        String functionSignature = isRuntimeFunction ? "RuntimeFunction: " : "ModuleFunction: ";
        functionSignature += returnTypeName + " " + functionName + " (" + parameterNames + ")";

        return functionSignature;
    }

    RemoteFunctionReturn makeRemoteFunctionReturn(RemoteFunctionRunnableResult result, RemoteFunctionRequest executionRequest)
    {
        RemoteFunctionReturn.Builder remoteFunctionReturn = RemoteFunctionReturn.newBuilder();
        RemoteFunctionIdentifier remoteFunctionIdentifier = executionRequest.getRemoteFunctionIdentifier();

        remoteFunctionReturn.setExecutionStatus(result.getStatus());
        remoteFunctionReturn.setRemoteFunctionIdentifier(remoteFunctionIdentifier);
        remoteFunctionReturn.setRemoteFutureIdentifier(executionRequest.getRemoteFutureIdentifier());


        return remoteFunctionReturn.build();
    }

    DataPackage makeRPCResponsePackage(RemoteFunctionRunnableResult result, DataPackage rpcRequest)
    {
        RemoteFunctionRequest executionRequest = rpcRequest.getControlVal().getRemoteFunctionRequest();

        DataPackage.Builder responseBuilder = DataPackage.newBuilder();

        responseBuilder.setSourceModule(rpcRequest.getTargetModule());
        responseBuilder.setTargetModule(rpcRequest.getSourceModule());


        ControlPackage.Builder ctrlPackage = ControlPackage.newBuilder();
        ctrlPackage.setCtrlType(CtrlType.CTRL_REMOTE_FUNCTION_RESPONSE);
        ctrlPackage.setRemoteFunctionReturn(makeRemoteFunctionReturn(result, executionRequest));

        ctrlPackage.setRuntime(Runtime.RUNTIME_JAVA);


        responseBuilder.setControlVal(ctrlPackage.build());

        DataPackage responsePackage = responseBuilder.build();

        Object returnValue = result.getReturnValue();
        if(returnValue != null)
        {
            Mutator<?> mutator = TypeMapping.getMutator(new DataType(this.returnType));
            responsePackage = mutator.setPackagePayloadFromObject(responsePackage, returnValue);
        }

        return responsePackage;
    }
}

File: dispatch/dart/claid/lib/RemoteFunction/RemoteFunctionRunnable.dart
/***************************************************************************
* Copyright (C) 2023 ETH Zurich
* CLAID: Closing the Loop on AI & Data Collection (https://claid.ethz.ch)
* Core AI & Digital Biomarker, Acoustic and Inflammatory Biomarkers (ADAMMA)
* Centre for Digital Health Interventions (c4dhi.org)
* 
* Authors: Patrick Langer
* 
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* 
*         http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
***************************************************************************/

import 'dart:async';
import 'package:claid/logger/Logger.dart';
import 'package:claid/generated/claidservice.pb.dart';
import 'package:claid/module/type_mapping.dart';
import 'RemoteFunctionRunnableResult.dart';


class RemoteFunctionRunnable<T>
{
  Function _function;
  T _returnType;
  List<dynamic> _parameterTypes; 

  RemoteFunctionRunnable(
      this._function, this._returnType, this._parameterTypes) 
  {

  }

  DataPackage executeRemoteFunctionRequest(DataPackage rpcRequest) 
  {
    RemoteFunctionRunnableResult<T> status;

    RemoteFunctionRequest executionRequest =
        rpcRequest.controlVal.remoteFunctionRequest;

    RemoteFunctionIdentifier remoteFunctionIdentifier =
        executionRequest.remoteFunctionIdentifier;

    int payloadsSize = executionRequest.parameterPayloads.length;

    if (payloadsSize != this._parameterTypes.length) 
    {
      Logger.logError(
          'Failed to execute RemoteFunctionRunnable "${getFunctionSignature(remoteFunctionIdentifier, executionRequest)}". Number of parameters do not match. Function expected ${this._parameterTypes.length} parameters, but was executed with $payloadsSize');

      status = RemoteFunctionRunnableResult.makeFailedResult(
          RemoteFunctionStatus.FAILED_INVALID_NUMBER_OF_PARAMETERS);
      return makeRPCResponsePackage(status, rpcRequest);
    }

    List<dynamic> parameters = [];
    for (int i = 0; i < payloadsSize; i++) {
      Mutator mutator = TypeMapping().getMutator(this._parameterTypes[i]);

      DataPackage tmpPackage = DataPackage();
      tmpPackage.payload = executionRequest.parameterPayloads[i];

      
      dynamic data = mutator.getter(tmpPackage);
      parameters.add(data);

      if (parameters[i].runtimeType != this._parameterTypes[i].runtimeType) 
      {
        Logger.logError(
            'Failed to execute RemoteFunctionRunnable "${getFunctionSignature(remoteFunctionIdentifier, executionRequest)}". Parameter object $i is of type "${parameters[i].runtimeType}", but expected type "${this._parameterTypes[i]}".');
        status = RemoteFunctionRunnableResult.makeFailedResult(
            RemoteFunctionStatus.FAILED_MISMATCHING_PARAMETERS);
        return makeRPCResponsePackage(status, rpcRequest);
      }
    }

    status = executeFunction(parameters);

    return makeRPCResponsePackage(status, rpcRequest);
  }

  RemoteFunctionRunnableResult<T> executeFunction(List<dynamic> parameters) 
  {
      T result = Function.apply(this._function, parameters);
      return RemoteFunctionRunnableResult.makeSuccessfulResult(result);
  }

  String getFunctionSignature(
      RemoteFunctionIdentifier remoteFunctionIdentifier,
      RemoteFunctionRequest remoteFunctionRequest) 
    {
    String returnTypeName = T == _getType<void>() ? 'void' : this._returnType.runtimeType.toString();

    bool isRuntimeFunction = remoteFunctionIdentifier.hasRuntime();

    String parameterNames =
        this._parameterTypes.isNotEmpty ? this._parameterTypes[0].runtimeType.toString() : '';

    for (int i = 1; i < this._parameterTypes.length; i++) {
      parameterNames += ', ' + this._parameterTypes[i].runtimeType.toString();
    }

  String functionName = remoteFunctionIdentifier.functionName;

    String functionSignature =
        isRuntimeFunction ? 'RuntimeFunction: ' : 'ModuleFunction: ';
    functionSignature +=
        '$returnTypeName $functionName ($parameterNames)';

    return functionSignature;
  }

  static RemoteFunctionReturn makeRemoteFunctionReturn(
      RemoteFunctionStatus status, RemoteFunctionRequest executionRequest) {
    RemoteFunctionReturn remoteFunctionReturn = RemoteFunctionReturn();

    RemoteFunctionIdentifier remoteFunctionIdentifier = executionRequest.remoteFunctionIdentifier;

    remoteFunctionReturn.executionStatus = status;
    remoteFunctionReturn.remoteFunctionIdentifier = remoteFunctionIdentifier;
    remoteFunctionReturn.remoteFutureIdentifier =
        executionRequest.remoteFutureIdentifier;

    return remoteFunctionReturn;
  }
  
  static DataPackage prepareResponsePackage(RemoteFunctionStatus status, DataPackage rpcRequest)
  {
    RemoteFunctionRequest executionRequest =
        rpcRequest.controlVal.remoteFunctionRequest;

    DataPackage responsePackage = DataPackage();

    responsePackage.sourceModule = rpcRequest.targetModule;
    responsePackage.targetModule = rpcRequest.sourceModule;

    ControlPackage ctrlPackage = ControlPackage();
    ctrlPackage.ctrlType = CtrlType.CTRL_REMOTE_FUNCTION_RESPONSE;
    ctrlPackage.remoteFunctionReturn =
        RemoteFunctionRunnable.makeRemoteFunctionReturn(status, executionRequest);

    // Send back to the runtime that made the RPC request.
    ctrlPackage.runtime = rpcRequest.controlVal.runtime;

    responsePackage.controlVal = ctrlPackage;

    return responsePackage;
  }

  DataPackage makeRPCResponsePackage(
      RemoteFunctionRunnableResult<T> result, DataPackage rpcRequest) {
    
    DataPackage responsePackage = prepareResponsePackage(result.getStatus(), rpcRequest);

    T? returnValue = result.getReturnValue();
    if(returnValue != null)
    {
      print("Return type " + this._returnType.runtimeType.toString());
        Mutator<T> mutator = TypeMapping().getMutator(this._returnType);
        mutator.setter(responsePackage, returnValue);
    }

    return responsePackage;
  }

    static bool isDataTypeSupported(dynamic dataTypeExample)
    {
      // Leads to compile error or exception if data type is not supported
      TypeMapping().getMutator(dataTypeExample);
      return true;
    }

    Type _getType<T>() => T;

}
